component-runtime-impl/src/main/java/org/talend/sdk/component/runtime/record/SchemaImpl.java

```diff
@@ -22,46 +22,62 @@
 import java.util.LinkedHashMap;
 import java.util.List;
 import java.util.Map;
+import java.util.stream.Stream;
 
 import javax.json.bind.annotation.JsonbTransient;
 
 import org.talend.sdk.component.api.record.Schema;
 
-import lombok.AllArgsConstructor;
 import lombok.Data;
-import lombok.NoArgsConstructor;
+import lombok.EqualsAndHashCode;
+import lombok.Getter;
 
-@Data
-@NoArgsConstructor
-@AllArgsConstructor
+@EqualsAndHashCode
 public class SchemaImpl implements Schema {
 
-    private Type type;
+    @Getter
+    private final Type type;
 
-    private Schema elementSchema;
+    @Getter
+    private final Schema elementSchema;
 
-    private List<Entry> entries;
+    @Getter
+    private final List<Entry> entries;
 
+    @JsonbTransient
+    private final List<Entry> metadataEntries;
+
+    @Getter
     private Map<String, String> props = new LinkedHashMap<>(0);
 
-    public SchemaImpl(final Type type, final Schema schema, final List<Entry> entries) {
-        this.type = type;
-        elementSchema = schema;
-        this.entries = entries;
+    SchemaImpl(final SchemaImpl.BuilderImpl builder) {
+        this.type = builder.type;
+        this.elementSchema = builder.elementSchema;
+        this.entries = unmodifiableList(builder.entries);
+        this.metadataEntries = unmodifiableList(builder.metadataEntries);
+        this.props = builder.props;
     }
 
     @Override
     public String getProp(final String property) {
         return props.get(property);
     }
 
+    @Override
+    @JsonbTransient
+    public Stream<Entry> getAllEntries() {
+        return Stream.concat(this.metadataEntries.stream(), this.entries.stream());
```

Comment:
here we may loose the ordering, no?

Reply:
No;
Collection.stream() method transform collection to stream with boolean parallel parameter to false.
So, concat will also have parallel to false; so order will be preserved.

Reply:
With the following code, I created some entries in a certain order...
```
final Schema schema = new BuilderImpl() //
                .withType(Type.RECORD) //
                .withEntry(newEntry("field1", "field1", Type.STRING, true, 5, "Comment"))
                .withEntry(new EntryImpl.BuilderImpl().withName("record_id").withMetadata(true).withType(Type.INT).withProp("method","FIFO").build())
                .withEntry(new EntryImpl.BuilderImpl().withName("field2").withMetadata(true).withType(Type.STRING).build())
                .build();
        final RecordImpl.BuilderImpl builder = new RecordImpl.BuilderImpl(schema);
        schema.getAllEntries().forEach(e-> System.out.println(e));
``` 
which outputs
 ```
SchemaImpl.EntryImpl(name=record_id, rawName=null, type=INT, nullable=false, metadata=true, defaultValue=null, elementSchema=null, comment=null, props={method=FIFO})
SchemaImpl.EntryImpl(name=field2, rawName=null, type=STRING, nullable=false, metadata=true, defaultValue=null, elementSchema=null, comment=null, props={})
SchemaImpl.EntryImpl(name=field1, rawName=field1, type=STRING, nullable=true, metadata=false, defaultValue=5, elementSchema=null, comment=Comment, props={})
```
So, order is not preserved.

Reply:
yes, metadata fields come before ... then classic field. 
Order is preserved inside it.
Is global ordering important ?

Reply:
Thinking so! 
Sorry, I stick to my initial idea... metadata should be treated as a property in schema like the nullable one.